### PR TITLE
Added OW_USE_FAKE_RTC and OW_ALTERED_TIME_RATIO

### PR DESCRIPTION
**Pull request by pkmnsnfrn**
# Description
Gif of the wallclock moving forward one minute for every second in the gif

* Created `OW_USE_FAKE_RTC`, which allows users to control the flow of time using frames instead of the RTC.
* Created `OW_ALTERED_TIME_RATIO`, which allows users to control the speed at which time flows within the game.

# Usage

## `OW_USE_FAKE_RTC`
In [`include/config/overworld.h`](https://github.com/PokemonSanFran/pokeemerald-expansion/blob/fakeRTC/include/config/overworld.h), developers must define if `OW_USE_FAKE_RTC` is `TRUE` or `FALSE`.

### == `TRUE`
The time within the game is determined by the number of frames that has passed since the clock was initalized. Every 60 frames, time increments by X seconds, as defined in `OW_ALTERED_TIME_RATIO`.

### == `FALSE`
The time within the game is determined by the reading of the RTC.

## `OW_ALTERED_TIME_RATIO`
In [`include/config/overworld.h`](https://github.com/PokemonSanFran/pokeemerald-expansion/blob/fakeRTC/include/config/overworld.h), developers must define which generation `OW_ALTERED_TIME_RATIO` should use.

This only works in conjuction with `OW_USE_FAKE_RTC`.

### == `GEN_8_PLA`
For every 60 frames, time in game moves forward 60 seconds.

### == `GEN_9`
For every 60 frames, time in game moves forward 20 seconds.

### != `GEN_8_PLA` && != `GEN_9`
For every 60 frames, time in game moves forward 1 second.

# Testing
## Clean Branch
You can recreate this branch by applying a patch or pulling the repo. From a clean version of expansion's upcoming, you can either:

### Patch
`wget https://raw.githubusercontent.com/PokemonSanFran/pokeemerald-expansion/fakeRTC_testing/fakeRTC.patch -O fakeRTC.patch ; git apply fakeRTC.patch ; rm fakeRTC.patch`

### Repo
`git remote add psf-expansion https://github.com/PokemonSanFran/pokeemerald-expansion/ ; git pull psf-expansion fakeRTC`

## Manual Tests
After replicating the branch, to recreate my testing environment, you can either directly download the debug script and config file, or manually create the changes.

### Download
| `OW_USE_FAKE_RTC`  |  `OW_ALTERED_TIME_RATIO` |  Command |
|---|---|---|
| `FALSE`  | `GEN_3` |  `wget https://raw.githubusercontent.com/PokemonSanFran/pokeemerald-expansion/fakeRTC_testing/include/config/f3.h -O include/config/overworld.h && wget https://raw.githubusercontent.com/PokemonSanFran/pokeemerald-expansion/fakeRTC_testing/data/scripts/debug.inc -O data/scripts/debug.inc`|
| `TRUE`  | `GEN_3` |  `wget https://raw.githubusercontent.com/PokemonSanFran/pokeemerald-expansion/fakeRTC_testing/include/config/t3.h -O include/config/overworld.h && wget https://raw.githubusercontent.com/PokemonSanFran/pokeemerald-expansion/fakeRTC_testing/data/scripts/debug.inc -O data/scripts/debug.inc`|
| `TRUE` | `GEN_8_PLA` |  `wget https://raw.githubusercontent.com/PokemonSanFran/pokeemerald-expansion/fakeRTC_testing/include/config/t8.h -O include/config/overworld.h && wget https://raw.githubusercontent.com/PokemonSanFran/pokeemerald-expansion/fakeRTC_testing/data/scripts/debug.inc -O data/scripts/debug.inc`|
| `TRUE`  | `GEN_9` |  `wget https://raw.githubusercontent.com/PokemonSanFran/pokeemerald-expansion/fakeRTC_testing/include/config/t9.h -O include/config/overworld.h && wget https://raw.githubusercontent.com/PokemonSanFran/pokeemerald-expansion/fakeRTC_testing/data/scripts/debug.inc -O data/scripts/debug.inc`|

### Manual Testing
* Change the config file and change `OW_USE_FAKE_RTC` and `OW_ALTERED_TIME_RATIO` to their desired values.
* Change [data/scripts/debug.inc](https://raw.githubusercontent.com/PokemonSanFran/pokeemerald-expansion/fakeRTC_testing/data/scripts/debug.inc) to match the one provided
* Compile the game
* Start a new save
* Run Script 1

## Verified Scenarios

All videos show:
* The player sets the clock to 15:00
* The player battles against `TRAINER_RONALD` and wins in 6 turns.
* A script prints the current time.

### `OW_USE_FAKE_RTC` == `FALSE`
https://github.com/rh-hideout/pokeemerald-expansion/assets/77138753/04f4fb0b-183c-4fea-b922-e03ff7aa448e

### `OW_USE_FAKE_RTC` == `TRUE` && `OW_ALTERED_TIME_RATIO == `GEN_3`
https://github.com/rh-hideout/pokeemerald-expansion/assets/77138753/77191e8e-7e4c-4b38-b352-f865567da5ac

### `OW_USE_FAKE_RTC` == `TRUE` && `OW_ALTERED_TIME_RATIO == `GEN_8_PLA`
https://github.com/rh-hideout/pokeemerald-expansion/assets/77138753/5796c741-2eb5-49e6-97dc-73aae68ec5fe

### `OW_USE_FAKE_RTC` == `TRUE` && `OW_ALTERED_TIME_RATIO == `GEN_9`
https://github.com/rh-hideout/pokeemerald-expansion/assets/77138753/3bb8a194-5087-4d2f-aa86-a9912e5ed041

# **People who collaborated with me in this PR**
All of this code was written by [Anon822](https://www.pokecommunity.com/members/anon822.699429/).

# Features this PR does NOT handle:
https://github.com/rh-hideout/pokeemerald-expansion/issues/4909

# Discord Contact Info
I am `pkmnsnfrn` on Discord.

### CODE REVIEW

File: include/config/overworld.h [hunk 1]
```diff
@@ -82,8 +81,16 @@
...
 #define OW_POPUP_BW_TIME_MODE      OW_POPUP_BW_TIME_NONE    // Determines what type of time is shown.
 #define OW_POPUP_BW_ALPHA_BLEND    FALSE                    // Enables alpha blending/transparency for the pop-ups. Mainly intended to be used with the black color option.
 
+// These generational defines only make a distinction for OW_ALTERED_TIME_RATIO
+#define GEN_8_PLA                       GEN_LATEST + 2
+
+//Time
+#define OW_TIMES_OF_DAY                 GEN_LATEST // Different generations have the times of day change at different times.
+#define OW_USE_FAKE_RTC                 FALSE      // When TRUE, seconds on the in-clock will only advance once every 60 playTimeVBlanks (every 60 frames).
+#define OW_ALTERED_TIME_RATIO           GEN_8_PLA  // In GEN_8_PLA, the time in game moves forward 60 seconds for every second in the RTC. In GEN_9, it is 20 seconds. This has no effect if OW_USE_FAKE_RTC is FALSE.
```
Comment (Bassoonian) [inline include/config/overworld.h:94]: Any reason this doesn't default to `GEN_LATEST`?

File: include/global.h [hunk 2]
```diff
@@ -181,6 +175,16 @@ struct Time
     /*0x04*/ s8 seconds;
 };
 
+
+struct SaveBlock3
+{
+#if OW_USE_FAKE_RTC
+	struct Time fakeRTC;
```
Comment (Bassoonian) [inline include/global.h:182]: spaces instead of tabs

File: src/fake_rtc.c [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+#include "string_util.h"
+#include "strings.h"
+#include "text.h"
+#include "rtc.h"
+#include "fake_rtc.h"
+
+struct Time *FakeRtc_GetCurrentTime(void)
+{
+#if OW_USE_FAKE_RTC
+	return &gSaveBlock3Ptr->fakeRTC;
```
Comment (Bassoonian) [inline src/fake_rtc.c:11]: spaces instead of tabs (for the entire file)

File: src/play_time.c [hunk 4]
```diff
@@ -45,6 +46,7 @@ void PlayTimeCounter_Update(void)
 
     gSaveBlock2Ptr->playTimeVBlanks = 0;
     gSaveBlock2Ptr->playTimeSeconds++;
+	FakeRtc_TickTimeForward();
```
Comment (Bassoonian) [inline src/play_time.c:49]: spaces instead of tabs

File: include/config/overworld.h (hunk 1 above)
Comment (pkmnsnfrn) [reply include/config/overworld.h]: https://github.com/rh-hideout/pokeemerald-expansion/pull/4910/commits/f1355ba536ce2843dc75196e853dd2dc5341f0b4

File: include/global.h (hunk 2 above)
Comment (pkmnsnfrn) [reply include/global.h]: https://github.com/rh-hideout/pokeemerald-expansion/pull/4910/commits/957dda9d1f078ae2fec18f1a7eaf881795225cb6

File: src/fake_rtc.c (hunk 3 above)
Comment (pkmnsnfrn) [reply src/fake_rtc.c]: https://github.com/rh-hideout/pokeemerald-expansion/pull/4910/commits/957dda9d1f078ae2fec18f1a7eaf881795225cb6

File: src/play_time.c (hunk 4 above)
Comment (pkmnsnfrn) [reply src/play_time.c]: https://github.com/rh-hideout/pokeemerald-expansion/pull/4910/commits/957dda9d1f078ae2fec18f1a7eaf881795225cb6

**Review by Bassoonian — APPROVED** [review]

**Pull request merged**